File: DSA/DSA_450/Python/14_DP/046_Count_Palindromic_Subsequence.py
"""
Problem: Count Palindromic Subsequence
URL: https://practice.geeksforgeeks.org/problems/count-palindromic-subsequences/1

Problem Statement:
Given a string str of length N, count the number of palindromic subsequences (not necessarily contiguous) in the string.

Sample Input/Output:
Input: "abcd"
Output: 4
Input: "aab"
Output: 4
"""


class Solution:
    def Count_Pal_Sub_Memo(self, s: str, i: int, j: int, dp: list[list[int]]) -> int:
        """
        Memoization approach
        Time Complexity: O(n^2)
        Space Complexity: O(n^2)
        """
        if i > j:
            return 0
        if i == j:
            return 1
        if dp[i][j] != -1:
            return dp[i][j]
        
        if s[i] == s[j]:
            dp[i][j] = (self.Count_Pal_Sub_Memo(s, i + 1, j, dp) + 
                       self.Count_Pal_Sub_Memo(s, i, j - 1, dp) + 1)
        else:
            dp[i][j] = (self.Count_Pal_Sub_Memo(s, i + 1, j, dp) + 
                       self.Count_Pal_Sub_Memo(s, i, j - 1, dp) - 
                       self.Count_Pal_Sub_Memo(s, i + 1, j - 1, dp))
        return dp[i][j]
    
    def Count_Pal_Sub_Tab(self, s: str) -> int:
        """
        Tabulation approach
        Time Complexity: O(n^2)
        Space Complexity: O(n^2)
        """
        n = len(s)
        dp = [[0] * n for _ in range(n)]
        
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if i == j:
                    dp[i][j] = 1
                elif s[i] == s[j]:
                    dp[i][j] = dp[i + 1][j] + dp[i][j - 1] + 1
                else:
                    dp[i][j] = dp[i + 1][j] + dp[i][j - 1] - dp[i + 1][j - 1]
        
        return dp[0][n - 1]


def Test_CountPalindromicSubsequence():
    solution = Solution()
    
    s1 = "abcd"
    dp1 = [[-1] * len(s1) for _ in range(len(s1))]
    result1 = solution.Count_Pal_Sub_Memo(s1, 0, len(s1) - 1, dp1)
    assert result1 == 4
    assert solution.Count_Pal_Sub_Tab(s1) == 4
    
    s2 = "aab"
    dp2 = [[-1] * len(s2) for _ in range(len(s2))]
    result2 = solution.Count_Pal_Sub_Memo(s2, 0, len(s2) - 1, dp2)
    assert result2 == 4
    assert solution.Count_Pal_Sub_Tab(s2) == 4


if __name__ == "__main__":
    Test_CountPalindromicSubsequence()

File: DSA/DSA_450/14_DP/046_Count_Palindromic_Subsequence.cpp
/*
Problem: Count Palindromic Subsequence
URL: https://practice.geeksforgeeks.org/problems/count-palindromic-subsequences/1

Problem Statement:
Given a string str of length N, count the number of palindromic subsequences (not necessarily contiguous) in the string.

Sample Input/Output:
Input: "abcd"
Output: 4
Input: "aab"
Output: 4
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long Count_Pal_Sub_Memo(string& s, int i, int j, vector<vector<long long>>& dp) {
        /*
        Memoization approach
        Time Complexity: O(n^2)
        Space Complexity: O(n^2)
        */
        if (i > j) return 0;
        if (i == j) return 1;
        if (dp[i][j] != -1) return dp[i][j];
        
        if (s[i] == s[j]) {
            dp[i][j] = Count_Pal_Sub_Memo(s, i + 1, j, dp) + 
                       Count_Pal_Sub_Memo(s, i, j - 1, dp) + 1;
        } else {
            dp[i][j] = Count_Pal_Sub_Memo(s, i + 1, j, dp) + 
                       Count_Pal_Sub_Memo(s, i, j - 1, dp) - 
                       Count_Pal_Sub_Memo(s, i + 1, j - 1, dp);
        }
        return dp[i][j];
    }
    
    long long Count_Pal_Sub_Tab(string& s) {
        /*
        Tabulation approach
        Time Complexity: O(n^2)
        Space Complexity: O(n^2)
        */
        int n = s.length();
        vector<vector<long long>> dp(n, vector<long long>(n, 0));
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (i == j) {
                    dp[i][j] = 1;
                } else if (s[i] == s[j]) {
                    dp[i][j] = dp[i + 1][j] + dp[i][j - 1] + 1;
                } else {
                    dp[i][j] = dp[i + 1][j] + dp[i][j - 1] - dp[i + 1][j - 1];
                }
            }
        }
        return dp[0][n - 1];
    }
};

void Test_Count_Pal_Sub() {
    Solution solution;
    
    string s1 = "abcd";
    vector<vector<long long>> dp1(s1.length(), vector<long long>(s1.length(), -1));
    long long result1 = solution.Count_Pal_Sub_Memo(s1, 0, s1.length() - 1, dp1);
    cout << "Memo: " << s1 << " -> " << result1 << endl;
    cout << "Tab: " << s1 << " -> " << solution.Count_Pal_Sub_Tab(s1) << endl;
    
    string s2 = "aab";
    vector<vector<long long>> dp2(s2.length(), vector<long long>(s2.length(), -1));
    long long result2 = solution.Count_Pal_Sub_Memo(s2, 0, s2.length() - 1, dp2);
    cout << "Memo: " << s2 << " -> " << result2 << endl;
    cout << "Tab: " << s2 << " -> " << solution.Count_Pal_Sub_Tab(s2) << endl;
}

int main() {
    Test_Count_Pal_Sub();
    return 0;
}
